### PR TITLE
Solve Chapter2.hs

### PR DESCRIPTION
**Pull request by fekir**
### Solutions for Chapter 2

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter2.hs [hunk 1]
```diff
@@ -336,7 +337,7 @@ from it!
 ghci> :l src/Chapter2.hs
 -}
 subList :: Int -> Int -> [a] -> [a]
-subList = error "subList: Not implemented!"
+subList b e l = if ( b < 0 || e < 0) then [] else take ( e - b + 1 ) (drop b l)
```
Comment (vrom911) [inline src/Chapter2.hs:340]: Your implementation is correct when the right bound is lower than the left. But I think you can implement it more efficiently by adding a separate case to your check :slightly_smiling_face:

File: src/Chapter2.hs [hunk 2]
```diff
@@ -500,7 +502,9 @@ True
 >>> isThird42 [42, 42, 0, 42]
 False
 -}
-isThird42 = error "isThird42: Not implemented!"
+isThird42 :: [Int] -> Bool
+isThird42 (_:(_:(42:_))) = True
```
Comment (vrom911) [inline src/Chapter2.hs:506]: This is correct. However, you usually don't write parenthesis explicitly in lists patterns, as the infix notation is fixed for that, so compiler usually understands it that way 🙂 It is like it is not needed to write parenthesis around multiplication as it has priority in calculation 🙂 
```suggestion
isThird42 (_:_:42:_) = True
```

File: src/Chapter2.hs [hunk 3]
```diff
@@ -605,7 +609,8 @@ Implement a function that duplicates each element of the list
 
 -}
 duplicate :: [a] -> [a]
-duplicate = error "duplicate: Not implemented!"
+duplicate [] = []
+duplicate (e:es) = [e, e] ++ duplicate es
```
Comment (vrom911) [inline src/Chapter2.hs:613]: Usually, adding elements to the end of the list is costly. Your implementation is correct :+1: But you can try to implement this function more efficiently by using `:` instead `++`. If you can think, how to avoid adding elements to the end of the list, and instead add to the beginning, but get the same result, that would be amazing!

In general, it's always a good idea to try to replace `l ++ [x]` with `x : l` whenever possible, though, in some cases it might not be that trivial to achieve the same result. Remember the analogy with trains :bullettrain_side: :train: :train: :train:


It is more efficient to construct lists using the cons operator directly without introducing intermediate values.
```suggestion
duplicate (x:xs) = x : x : duplicate xs
```

File: src/Chapter2.hs [hunk 4]
```diff
@@ -620,7 +625,10 @@ Write a function that takes elements of a list only on even positions.
 >>> takeEven [2, 1, 3, 5, 4]
 [2,3,4]
 -}
-takeEven = error "takeEven: Not implemented!"
+takeEven :: [a] -> [a]
+takeEven [] = []
+takeEven (x:(_:y)) = x : takeEven(y)
```
Comment (vrom911) [inline src/Chapter2.hs:630]: Yeah, this is wonderful pattern matching skills.

Note that you don't need parenthesis. Also, functions and arguments in Haskell are always separated by spaces, even when in parenthesis 🙂  `takeEven (...)`.
```suggestion
takeEven (x:_:y) = x : takeEven y
```

File: src/Chapter2.hs [hunk 5]
```diff
@@ -727,7 +735,9 @@ value of the element itself
 🕯 HINT: Use combination of 'map' and 'replicate'
 -}
 smartReplicate :: [Int] -> [Int]
-smartReplicate l = error "smartReplicate: Not implemented!"
+smartReplicate [] = []
+smartReplicate (x:xs) = replicate x x ++ smartReplicate xs
```
Comment (vrom911) [inline src/Chapter2.hs:739]: This is right! 💯 

You can notice that what is happening is you walk through each element of the list, change it to something else (by using `replicate` function) and then concatenate all of the results into one list.

In Haskell, there is the `map` function that does exactly the first part of it (walk through each element of the list, change it to something else ).

After using `map` you will get a list of lists. To flatten such a list, you can use the `concat` function:

```haskell
ghci> concat [[3, 1, 2], [4, 0, 1]]
[3,1,2,4,0,1]
```

This pattern of using `map` and `concat` is very common, so the standard library has even a standard function `concatMap` which is a combination of `map` and `concat` (as the name suggests).

File: src/Chapter2.hs [hunk 6]
```diff
@@ -740,7 +750,8 @@ the list with only those lists that contain a passed element.
 
 🕯 HINT: Use the 'elem' function to check whether an element belongs to a list
 -}
-contains = error "contains: Not implemented!"
+contains :: Int -> [[Int]] -> [[Int]]
+contains n xxs = filter (\xs -> elem n xs) xxs
```
Comment (vrom911) [inline src/Chapter2.hs:754]: Now, after you've mastered eta-reduction, you can apply such a technique to other functions and internal lambda function as well  🙂 
But your implementation is already great 💯

File: src/Chapter2.hs [hunk 7]
```diff
@@ -780,13 +791,15 @@ Let's now try to eta-reduce some of the functions and ensure that we
...
 
 -- TODO: type ;)
-listElementsLessThan x l = filter (< x) l
+listElementsLessThan :: Int -> [Int] -> [Int]
+listElementsLessThan x = filter (< x)
 
 -- Can you eta-reduce this one???
-pairMul xs ys = zipWith (*) xs ys
+pairMul :: [Int] -> [Int]-> [Int]
+pairMul = zipWith (*)
```
Comment (vrom911) [inline src/Chapter2.hs:802]: Eta-reduction award 🏆

File: src/Chapter2.hs [hunk 8]
```diff
@@ -841,7 +854,8 @@ list.
 
 🕯 HINT: Use the 'cycle' function
 -}
-rotate = error "rotate: Not implemented!"
+rotate :: Int -> [Int] -> [Int]
+rotate n l = if (n<0) then [] else take (length l) (drop n (cycle l))
```
Comment (vrom911) [inline src/Chapter2.hs:858]: Some spaces suggestions 🙂 👾 
```suggestion
rotate n l = if (n < 0) then [] else take (length l) (drop n (cycle l))
```

File: src/Chapter2.hs [hunk 9]
```diff
@@ -857,7 +871,9 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
+rewind :: [a] -> [a]
+rewind [] = []
+rewind (x:xs) = rewind xs ++ [x]
```
Comment (vrom911) [inline src/Chapter2.hs:876]: Your solution is correct! However, it is slow. In lists, it is quite slow to add anything at the end of the list. That is why it is always better to rewrite it with the `:` cons. Remember the explanation with trains? 🚂 🚋 🚋 

That is why a more efficient solution is with the accumulator and the recursive function that will do the addition at the start of the list which is instant!

You can read a bit more about the `go` pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#recursive-go

**Review by vrom911 — APPROVED** [review]
Fantastic! Chapter Two is done ✔️ Keep up the great work 👏🏼

File: src/Chapter2.hs (hunk 1 above)
Comment (fekir) [reply src/Chapter2.hs:340]: Thank you,  do you mean something like
````
subList b e l = if ( b < 0 || e < 0 || e <= b ) then [] else take ( e - b + 1 ) (drop b l)
````

doesn't special-casing make it generally more difficult to understand the expression?
I thought, from a performance perspective, that the lazyness would make it irrelevant.

File: src/Chapter2.hs (hunk 2 above)
Comment (fekir) [reply src/Chapter2.hs:506]: It seemed to me that `isThird42 (_:_:42:_) = True` did not work, but apparently I must have done something else wrong when I tried it, thanks

File: src/Chapter2.hs (hunk 7 above)
Comment (fekir) [reply src/Chapter2.hs:802]: thanks, even if I can't say I find it easier to read (yet)

File: src/Chapter2.hs (hunk 3 above)
Comment (fekir) [reply src/Chapter2.hs:613]: ok, thank you

File: src/Chapter2.hs (hunk 5 above)
Comment (fekir) [reply src/Chapter2.hs:739]: Ahhh, now I got it, thanks

File: src/Chapter2.hs (hunk 9 above)
Comment (fekir) [reply src/Chapter2.hs:876]: ok, implemented a solution with the `go` pattern.

Is there a rule of thumb for deciding to use the `go`-pattern, thus creating temporaries, or using in this case `++`?

Isn't it effectively the same at a higher level perspective?

File: src/Chapter2.hs (hunk 1 above)
Comment (vrom911) [reply src/Chapter2.hs:340]: Yes, but `e < b` 🙂 
If you think that it looks too wordy, you can rewrite it to the guard casing, or just split if then else into 3 lines 🙂

File: src/Chapter2.hs (hunk 7 above)
Comment (vrom911) [reply src/Chapter2.hs:802]: Yeah, totally get you there! It is just the exercise, you could keep the arguments for readability and use `eta-reduction only when it would bring some benefits, or help to maintain or understand code 👍🏼 Good point!

File: src/Chapter2.hs (hunk 9 above)
Comment (vrom911) [reply src/Chapter2.hs]: The rule of thumb is to always use `:` instead of `++`. The thing is, that it is not always possible. But it is something to keep in mind while working with lists in Haskell.

Go pattern is quite a standard thing. There are also a few other tricks to make the solutions faster (like making the evaluation of the accumulator at each step, instead of laziness in there), so it is usually the good way to go 🙂

File: src/Chapter2.hs (hunk 1 above)
Comment (fekir) [reply src/Chapter2.hs:340]: Yes, `e<b`, or maybe better: `e-b+1 <= 0`

But that leaves my question unanswered: does it make really a difference performance-wise?
As far as I've understood, and I'm probably wrong, after the call to `take` with `e-b+1` no action is made on the list, thous should not make a difference.


My comment was not about the wordiness, but about thinking about special cases in the code.
If from a performance perspective there are no differences, I do not see the advantage of special-casing whats not a special case for `take`.

Comment (vrom911) [reply src/Chapter2.hs:340]: It makes the difference as with the special case you safe a few additional function calls which could be priceless.
Sorry if I did not answer your question explicitly. It would help if you would be more precise in your questions. "doesn't special-casing make it generally more difficult to understand the expression?" still seems to me about the style and formating and not about the performance at all. I also wrote about the performance in my initial comment.

Comment (vrom911) [reply src/Chapter2.hs:340]: And I don't agree that `e-b+1 <= 0` is better than `e < b`.

Comment (fekir) [reply src/Chapter2.hs:340]: I thought `e-b+1 <= 0` would be better as `e-b+1` it is the same expression that appears as parameter to `take`, so it's more obvious why we are testing it.

> It makes the difference as with the special case you safe a few additional function calls which could be priceless.

Ok, I thought function calls where not costly, unless they invoke system calls (like memory allocation, file access), networks resources or interprocess communication.

According to the behavior I've observed, I believe a possible implementation to `take` could be similar to

````
take :: Int -> [a] -> [a]
take _ []              =  []
take n (x:xs)          =  if n <= 0 : then [] else x : take (n-1) xs
````

thus to make no difference, or at least be less efficient as we are duplicating the test if the parameter to `take` is positive or not.


> It would help if you would be more precise in your questions. "doesn't special-casing make it generally more difficult to understand the expression?"

I simply meant, that by reading the code, one asks itself why `e<b` needs to be treated separately as the code without the separate test behaves the same way.
One could think that `take` does not handle that case (or not always, which AFAIK is not the case), and implement similar checks in front of every call to `take`.
